File: objc/ATValidations.m
//
//  ATValidations.m
//  Azure Talon
//
//  Created by Kenneth Ballenegger on 12/10/12.
//
//

#import "ATValidations.h"


@interface ATVPredicate : NSObject
@property (copy) BOOL(^block)(id object, NSError **error);
@end

NSString *const kATVPredicateError = @"ATVPredicateError";





BOOL ATVAssert(BOOL assertion, NSString *errorMessage, NSError **error) {
    if (!assertion && error) {
        *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{@"error": errorMessage}];
    }
    return assertion;
}





// matches block is the master definition, all others simply use this definition
ATVPredicate *ATVMatchesBlock(BOOL(^block)(id object, NSError **error)) {
    ATVPredicate *type = [[ATVPredicate alloc] init];
    type.block = block;
    return [type autorelease];
}

ATVPredicate *ATVUnionA(id<NSFastEnumeration> predicates) {

    return ATVMatchesBlock(^BOOL(id object, NSError **error) {

        NSError *failure = nil;
        for (ATVPredicate *predicate in predicates) {
            if (!predicate.block(object, &failure)) {
                if (error) {
                    *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{
                               @"error": @"union predicate failed",
                               @"failure": failure.userInfo
                               }];
                }
                return NO;
            }
        }
        return YES;
    });
}

ATVPredicate *ATVOptionA(id<NSFastEnumeration> predicates) {
    
    return ATVMatchesBlock(^BOOL(id object, NSError **error) {

        NSMutableArray *failures = [NSMutableArray array];
        for (ATVPredicate *predicate in predicates) {
            NSError *failure = nil;
            if (predicate.block(object, &failure)) {
                return YES;
            } else {
                [failures addObject:failure.userInfo];
            }
        }
        
        if (error) {
            *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{
                       @"error": @"option predicate failed",
                       @"failures": failures
                       }];
        }
        return NO;
    });
}

ATVPredicate *ATVEqual(id value) {
    return ATVMatchesBlock(^BOOL(id object, NSError **error) {
        return ATVAssert([object isEqual:value], [NSString stringWithFormat:@"must be equal: %@ != %@", object, value], error);
    });
}

ATVPredicate *ATVExists() {
    static ATVPredicate *instance = nil;
    if (!instance) {
        instance = [ATVMatchesBlock(^BOOL(id object, NSError **error) {
            return ATVAssert(object != nil, @"must exist", error);
        }) retain];
    }
    return instance;
}

ATVPredicate *ATVNull() {
    static ATVPredicate *instance = nil;
    if (!instance) {
        instance = [ATVMatchesBlock(^BOOL(id object, NSError **error) {
            return ATVAssert(object == nil, @"must be null", error);
        }) retain];
    }
    return instance;
}

ATVPredicate *ATVNullOr(ATVPredicate *predicate) {
    return ATVOption(ATVNull(), predicate);
}


ATVPredicate *ATVInSetA(id<NSFastEnumeration> values) {

    return ATVMatchesBlock(^BOOL(id object, NSError **error) {
        
        for (id value in values) {
            if ([object isEqual:value]) {
                return YES;
            }
        }
        
        if (error) {
            *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{
                       @"error": @"must be in set of values provided",
                       @"options": values
                       }];
        }
        return NO;
    });
}

ATVPredicate *ATVArrayOf(ATVPredicate *predicate) {
    return ATVMatchesBlock(^BOOL(id object, NSError **error) {
        if (![object isKindOfClass:[NSArray class]]) {
            if (error) {
                *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{
                                @"error": @"expecting array"
                           }];
            }
            return NO;
        }
        __block BOOL valid = YES;
        NSMutableDictionary *errors = [NSMutableDictionary dictionary];
        [(NSArray *)object enumerateObjectsUsingBlock:^(id member, NSUInteger idx, BOOL *stop) {
            NSError *error;
            BOOL memberValid = predicate.block(member, &error);
            if (!memberValid && error) {
                errors[@(idx)] = error.userInfo;
            }
            valid *= memberValid;
        }];
        
        if (error && !valid) {
            *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{
                           @"error": @"array children invalid",
                           @"failures": errors
                       }];
        }
        
        return valid;
    });
}

ATVPredicate *ATVDictionary(NSDictionary *mask, BOOL allowExtra) {
    return ATVMatchesBlock(^BOOL(id object, NSError **error) {
        if (![object isKindOfClass:[NSDictionary class]]) {
            if (error) {
                *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{
                               @"error": @"expecting dictionary"
                           }];
            }
            return NO;
        }
        __block BOOL valid = YES;
        NSMutableDictionary *errors = [NSMutableDictionary dictionary];
        void(^eachObject)(id key, id member, BOOL *stop) = ^(id key, id member, BOOL *stop) {
            
            ATVPredicate *predicate = mask[key];
            if (predicate) assert([predicate isKindOfClass:[ATVPredicate class]]);
            
            BOOL memberValid = YES;
            if (!allowExtra && !predicate) {
                memberValid = NO;
                errors[key] = @{@"error" : @"extra key not allowed", @"key": key};
            } else if (predicate) {
                NSError *error;
                memberValid = predicate.block(member, &error);
                if (!memberValid && error) {
                    errors[key] = error.userInfo;
                }
            }
            valid *= memberValid;
            
        };
        
        [(NSDictionary *)object enumerateKeysAndObjectsUsingBlock:eachObject];
        [(NSDictionary *)mask enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
            if (!((NSDictionary *)object)[key]) {
                eachObject(key, nil, nil);
            }
        }];
        
        if (error && !valid) {
            *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{
                           @"error": @"dictionary children invalid",
                           @"failures": errors
                       }];
        }
        
        return valid;
    });
}

ATVPredicate *ATVNumber(void) {
    static ATVPredicate *instance = nil;
    if (!instance)
        instance = [ATVInstanceOf([NSNumber class]) retain];
    return instance;
}

ATVPredicate *ATVString(void) {
    static ATVPredicate *instance = nil;
    if (!instance)
        instance = [ATVInstanceOf([NSString class]) retain];
    return instance;
}

ATVPredicate *ATVInstanceOf(Class klass) {
    return ATVMatchesBlock(^BOOL(id object, NSError **error) {
        if (![object isKindOfClass:klass]) {
            if (error) {
                *error = [NSError errorWithDomain:kATVPredicateError code:1 userInfo:@{
                               @"error": [NSString stringWithFormat:@"expecting instance of class %@", NSStringFromClass(klass)]
                           }];
            }
            return NO;
        } else {
            return YES;
        }
    });

}



@implementation ATVPredicate
- (void)dealloc {
    self.block = nil;
    [super dealloc];
}
@end


@implementation NSObject (ATValidations)

- (BOOL)matchesATVPredicate:(ATVPredicate *)type error:(NSError **)error {
    return type.block(self, error);
}

@end

@implementation NSDictionary (ATValidations)

- (BOOL)matchesMask:(NSDictionary *)mask allowExtraKeys:(BOOL)allowExtra error:(NSError **)error {
    return [self matchesATVPredicate:ATVDictionary(mask, allowExtra) error:error];
}

@end

File: ruby/lib/at-validations.rb
require File.expand_path("#{File.dirname(__FILE__)}/at-validations/version.rb")


# We're adding (monkey-patching) a method to Object and to Hash.
#
# Hash gets this convenience method:
#
#   {hello: :world}.matches_mask({
#     _id: atv_string,
#     status: atv_union(
#               atv_number,
#               atv_block {|e| e >= 200 && e < 300 })
#   })
#

class Hash
  def matches_mask(mask, opts = {})
    ATValidations::Predicates.atv_hash(mask, opts).call(self)
  end
end

class Object
  def matches_predicate(predicate)
    predicate.call(self)
  end
end


module ATValidations

  module Predicates
    def atv_block(&b)
      b
    end

    def atv_hash(mask, opts = {})
      atv_union(
        atv_instance_of(Hash),
        atv_block do |e|
          errors = {}
          mask.each do |k, v|
            r = v.call(e[k])
            errors[k] = r unless true == r
          end
          if false == opts[:allow_extra] && (extra = (e.keys - mask.keys)).count > 0
            extra.each {|k| errors[k] = 'is not present in predicate' }
          end

          errors.empty? || Error.new(:error => 'must match hash predicate', :failures => errors)
        end
      )
    end

    def atv_array_of(predicate)
      atv_union(
        atv_instance_of(Array),
        atv_block do |e|
          errs = {}
          e.each_index {|i| err = predicate.call(e); errs[i] = err unless true == err }
          errs.count == 0 || Error.new(:error => 'array contains elements which do not match predicate', :failures => errs)
        end
      )
    end

    def atv_union(*predicates)
      atv_block do |e|
        err = nil
        nil == predicates.find {|p| true != (err = p.call(e)) } ||
          Error.new(:error => 'must match all predicate in union', :failure => err)
      end
    end

    def atv_option(*predicates)
      atv_block do |e|
        errs = []
        nil != predicates.find {|p| true == (errs << p.call(e)).last } ||
          Error.new(:error => 'must match at least one predicate in option', :failure => errs)
      end
    end

    def atv_in_set(*set)
      atv_block do |e|
        set.include?(e) || Error.new(:error => "must be in set #{set}")
      end
    end

    def atv_equal(value)
      atv_block do |e|
        e == value || Error.new(:error => "must be equal to #{value}")
      end
    end

    def atv_instance_of(klass)
      atv_block do |e|
        e.is_a?(klass) || Error.new(:error => "must be a #{klass}")
      end
    end

    def atv_numeric
      atv_instance_of(Numeric)
    end

    def atv_string
      atv_instance_of(String)
    end

    def atv_symbol
      atv_instance_of(Symbol)
    end

    def atv_nil_or(predicate)
      atv_block do |e|
        nil == e || predicate.call(e)
      end
    end
  end

  class Error < StandardError
    def initialize(info = {})
      @info = info
    end
    def to_s
      @info.to_s
    end
  end
end
